src/derivative_operators/concretization.jl

```diff
@@ -115,45 +118,74 @@ function SparseArrays.SparseMatrixCSC(Q::AffineBC{T}, N::Int) where {T}
+    # conditions. If these coefficients are zero (Dirichlet0BC), then
+    # the proper bandwidths are (l,u) = (1,-1).    
+    l = max(count(!iszero, a_r)+1, 1)
+    u = max(count(!iszero, a_l)-1, -1)
+
+    Q_l = BandedMatrix((-1 => ones(T,N),), (N+2,N), (l, u))
+    for (j,e) ∈ enumerate(a_l)
+        BandedMatrices.inbands_setindex!(Q_l, e, 1, j)
+    end
+    for (j,e) ∈ enumerate(a_r)
+        BandedMatrices.inbands_setindex!(Q_l, e, N+2, N-length(a_r)+j)
+    end
     Q_b = [Q.b_l; zeros(T,N); Q.b_r]
-    return (Q_l, Q_b)
+    
+    Q_l, Q_b
 end
 
-function SparseArrays.sparse(Q::AffineBC{T}, N::Int) where {T}
-    SparseMatrixCSC(Q,N)
-end
+"""
+    sparse(Q::AffineBC, N)
 
-LinearAlgebra.Array(Q::PeriodicBC{T}, N::Int) where T = (Array([transpose(zeros(T, N-1)) one(T); Diagonal(ones(T,N)); one(T) transpose(zeros(T, N-1))]), zeros(T, N))
-SparseArrays.SparseMatrixCSC(Q::PeriodicBC{T}, N::Int) where T = ([transpose(zeros(T, N-1)) one(T); Diagonal(ones(T,N)); one(T) transpose(zeros(T, N-1))], zeros(T, N))
-SparseArrays.sparse(Q::PeriodicBC{T}, N::Int) where T = SparseMatrixCSC(Q,N)
-function BandedMatrices.BandedMatrix(Q::PeriodicBC{T}, N::Int) where T #Not reccomended!
-    Q_array = BandedMatrix{T}(Eye(N), (N-1, N-1))
-    Q_array[1, end] = one(T)
-    Q_array[1, 1] = zero(T)
-    Q_array[end, 1] = one(T)
-    Q_array[end, end] = zero(T)
-
-    return (Q_array, zeros(T, N))
-end
+Since affine boundary conditions are representable by banded matrices,
+that is the default sparse concretization; if you want a
+`SparseMatrixCSC`, use `SparseMatrixCSC(Q, N)` instead.
+"""
+SparseArrays.sparse(Q::AffineBC, N::Int) = BandedMatrix(Q,N)
 
-function LinearAlgebra.Array(Q::BoundaryPaddedVector)
-    return [Q.l; Q.u; Q.r]
-end
+# ** Periodic BCs
 
-function Base.convert(::Type{Array},A::AbstractBC{T}) where T
-    Array(A)
-end
+LinearAlgebra.Array(Q::PeriodicBC{T}, N::Int) where T =
+    ([transpose(zeros(T, N-1)) one(T)
+      Diagonal(ones(T,N))
+      one(T) transpose(zeros(T, N-1))],
+     zeros(T,N+2))
 
-function Base.convert(::Type{SparseMatrixCSC},A::AbstractBC{T}) where T
-    SparseMatrixCSC(A)
-end
+SparseArrays.SparseMatrixCSC(Q::PeriodicBC{T}, N::Int) where T =
+    (vcat(hcat(zeros(T, 1,N-1), one(T)),
+          Diagonal(ones(T,N)),
+          hcat(one(T), zeros(T, 1, N-1))),
+     zeros(T,N+2))
 
-function Base.convert(::Type{AbstractMatrix},A::AbstractBC{T}) where T
-    SparseMatrixCSC(A)
-end
+SparseArrays.sparse(Q::PeriodicBC{T}, N::Int) where T = SparseMatrixCSC(Q,N)
+
+BandedMatrices.BandedMatrix(::PeriodicBC, ::Int) =
+    throw(ArgumentError("Periodic boundary conditions should be concretized as sparse matrices"))
+
+LinearAlgebra.Array(Q::BoundaryPaddedVector) = [Q.l; Q.u; Q.r]
+
+Base.convert(::Type{Array},A::AbstractBC) = Array(A)
+Base.convert(::Type{SparseMatrixCSC},A::AbstractBC) = SparseMatrixCSC(A)
+Base.convert(::Type{AbstractMatrix},A::AbstractBC) = SparseMatrixCSC(A)

```

Comment:
IMO, this should act like `sparse` and give "the best choice"

Reply:
Well, none of these `convert` methods actually work at all, since the matrix constructors require a length argument `N` to concretize a boundary condition, and there is no possible default length.

Reply:
Then maybe it's best to just delete them?

Reply:
Will do.